Add animations module with location animators

KacperKluka opened the pull request:
I've created a new module that is meant to be released as a maven package. This module will contain all animation related code that will make it easier to have a good subscriber UI/UX out of the box. For now, I've moved there the animation related code that was previously in the subscriber example app.

It would be great if all reviewers could get a look at the [CoreLocationAnimator file](https://github.com/ably/ably-asset-tracking-android/blob/eaf2a00906540a30e4834076378f6cd5c5bf71ae/animation/src/main/java/com/ably/tracking/animation/CoreLocationAnimator.kt) as it should be reviewed but git considers that as a file rename instead of creating a new file :unamused:

ikbalkaya reviewed (approved):
Looks good to me 👍

QuintinWillison reviewed (changes requested):
The name `animation` feels too specific for where I reckon this module might end up going. I would like to suggest:

- rename `animation/` to `ui-sdk/` (module folder)
- rename module root package from `com.ably.tracking.animation` to `com.ably.tracking.ui`
- move all of your current API surface to package `com.ably.tracking.ui.animation`

I also wonder about the `Position` data class... I've not looked for lack of time available to me in this moment, but can't that be in `core-sdk/`? 🤔

QuintinWillison commented on animation/src/main/java/com/ably/tracking/animation/LocationAnimator.kt:
```diff
@@ -0,0 +1,64 @@
+     */
+    fun animateLocationUpdate(
+        locationUpdate: LocationUpdate,
+        expectedIntervalBetweenLocationUpdatesInMilliseconds: Long,
```
My preference has always been to include trailing commas in any language that supports this, as it makes future Git diffs much cleaner. The tools can catch up, or be configured. 😉

KacperKluka replied:
We have not yet experienced any issues with trailing commas so I will leave it as it is but thanks for the information, I didn't know that :grin:

KacperKluka commented on animation/src/main/java/com/ably/tracking/animation/LocationAnimator.kt:
```diff
@@ -0,0 +1,64 @@
+     * The flow of subsequent camera positions that helps to follow the animation progress from [positionsFlow].
+     * The position of the camera can be changed each time a new value arrives in this flow.
+     */
+    val cameraPositionsFlow: SharedFlow<Position>
```
Thanks, for some reason I've forgotten that `SharedFlow` is a subtype of the `Flow` :sweat_smile: https://github.com/ably/ably-asset-tracking-android/pull/645/commits/bbd5d333692f9b3833f349c69a5630dfe626ea1b

KacperKluka commented:
@QuintinWillison I've renamed the module to `ui-sdk` https://github.com/ably/ably-asset-tracking-android/pull/645/commits/e40154bfe43c33263a50d6cda4aece0f2df73fba and also moved the animation classes to the `ui.animation` package https://github.com/ably/ably-asset-tracking-android/pull/645/commits/f45ee6e28de965d17a4f1b99829cfa9d22c0178c

Regarding the `Position` class, it is a plain Kotlin class with primitive values but it is only used in the `LocationAnimator` that's why I have left it there :wink:

QuintinWillison commented on ui-sdk/build.gradle:
```diff
@@ -0,0 +1,23 @@
+}
+
+dependencies {
+    implementation project(':subscribing-sdk')
```
Sorry, I should have spotted this in my first-pass review of this... why is this here?

```
ably-asset-tracking-android % cd ui-sdk
ui-sdk % git grep -i subscriber
src/main/java/com/ably/tracking/ui/animation/LocationAnimator.kt: * Extension for the [com.ably.tracking.subscriber.Subscriber] that enables smooth location updates animation.
src/main/java/com/ably/tracking/ui/animation/LocationAnimator.kt:     * This should be called each time a new location update is received from the [com.ably.tracking.subscriber.Subscriber].
src/main/java/com/ably/tracking/ui/animation/LocationAnimator.kt:     * @param locationUpdate The newest location update received from the Subscriber SDK.
```

This is an unnecessary dependency. If anything, I would have assumed you would need an API/interface dependency on `':core-sdk'` for `LocationUpdate`. 🤔 

I can clearly see that you are referencing `Subscriber` in the comments my `grep` revealed, but that should not then require you to make it an implementation dependency because the code shows that not to be true.

KacperKluka replied:
Since this module is meant to be an extension for the `subscirber-sdk` module it felt like a good choice to make it dependent on the subscriber module :thinking: But from the technical point of view we can replace it with the `core-sdk` :wink:

QuintinWillison replied:
Let's stick to the concrete. Please change this. Thanks.

KacperKluka replied:
https://github.com/ably/ably-asset-tracking-android/pull/645/commits/d9c96a0f80c603edb2a77a5d54b084c3f7b16ca3
